data_management/rocrate.py

```diff
@@ -0,0 +1,661 @@
+
+    if crate_data_product is None:
+        crate_data_product = _get_local_data_product(
+            crate, data_product, registry_url, output
+        )
+
+    _add_licenses(crate, crate_data_product, data_product.object, registry_url)
+
+    _add_authors(
+        data_product.object.authors.all(),
+        crate,
+        crate_data_product,
+        registry_url,
+    )
+
+    return crate_data_product
+
+
+def _get_external_object(crate, data_product):
+    """
+    Create an RO Crate file entity for the given data product.
+
+    @param crate: the RO Crate object
+    @param data_product: a data_product from the DataProduct table
+
+    @return an RO Crate file entity representing the external object, may be None
+
+    """
+    # check for external object linked to the data product
+    try:
+        external_object = data_product.external_object
+    except (models.DataProduct.external_object.RelatedObjectDoesNotExist,):
+        return None
+
+    properties = {}
+    properties["name"] = external_object.title
+    properties["sdDatePublished"] = str(external_object.release_date)
+
+    if external_object.identifier:
+        source_loc = external_object.identifier
+    else:
+        source_loc = external_object.alternate_identifier
+
+    if external_object.description:
+        properties["description"] = external_object.description
+
+    if external_object.original_store:
+        properties["original_store"] = str(external_object.original_store)
+
+    crate_external_object = crate.add_file(source_loc, properties=properties)
+
+    return crate_external_object
+
+
+def _get_local_data_product(crate, data_product, registry_url, output):
+    """
+    Create an RO Crate file entity representing the data product.
+
+    @param crate_code_run: RO Crate entity representing the code run
+    @param data_product: a data_product from the DataProduct table
+    @param registry_url: a str containing the registry URL
+    @param output (bool): true if the data product is an output
+
+    @return an RO Crate file entity representing the data product
+
+    """
+    if (
+        data_product.object.storage_location.public is True
+        and len(str(data_product.object.storage_location).split(FILE)) > 1
+    ):
+        source_loc = str(data_product.object.storage_location).split(FILE)[1]
+
+        if output:
+            dest_path = f"outputs/{source_loc.split('/')[-1]}"
+        else:
+            dest_path = f"inputs/data/{source_loc.split('/')[-1]}"
+
+    else:
+        source_loc = f"{registry_url}api/storage_location/{data_product.object.storage_location.id}"
+        dest_path = None
+
+    encoding_format = _get_mime_type(data_product.object.file_type.extension)
+
+    properties = {
+        "name": data_product.name,
+        "encodingFormat": encoding_format,
+    }
+
+    if data_product.object.storage_location.hash is not None:
+        properties["hash"] = data_product.object.storage_location.hash
```

Comment:
"hash" is not a valid property. In the RO-Crate context, "File" is an alias for [MediaObject](https://schema.org/MediaObject), so [sha256](https://schema.org/sha256) can be used.

Reply:
updated in https://github.com/FAIRDataPipeline/data-registry/commit/eff083a65a8c2f1844c7934304a21405a46c61fd